src/ray/util/process.h

```diff
@@ -0,0 +1,94 @@
+
+// We only define operators required by the standard library (<, ==, hash).
+// We declare but avoid defining the rest so that they're not used by accident.
+
+namespace std {
+template <>
+struct hash<boost::process::child> : hash<boost::process::pid_t> {
+  typedef hash<boost::process::pid_t> base_type;
+  typedef boost::process::child argument_type;
+  typedef size_t result_type;
+  result_type operator()(const argument_type &value) const {
+    return this->base_type::operator()(value.id());
+  }
+};
+template <>
+struct equal_to<boost::process::child> : equal_to<boost::process::pid_t> {
+  typedef equal_to<boost::process::pid_t> base_type;
+  typedef boost::process::child argument_type;
+  typedef bool result_type;
+  result_type operator()(const argument_type &a, const argument_type &b) const {
+    return this->base_type::operator()(a.id(), b.id());
+  }
+};
+template <>
+struct greater<boost::process::child>;
+template <>
+struct greater_equal<boost::process::child>;
+template <>
+struct less<boost::process::child> : less<boost::process::pid_t> {
+  typedef less<boost::process::pid_t> base_type;
+  typedef boost::process::child argument_type;
+  typedef bool result_type;
+  result_type operator()(const argument_type &a, const argument_type &b) const {
+    return this->base_type::operator()(a.id(), b.id());
+  }
+};
+template <>
+struct less_equal<boost::process::child>;
+template <>
+struct not_equal_to<boost::process::child>;
+
+template <>
+struct hash<shared_ptr<boost::process::child>> : hash<boost::process::child> {
+  typedef hash<boost::process::child> base_type;
+  typedef shared_ptr<boost::process::child> argument_type;
+  typedef size_t result_type;
+  result_type operator()(const argument_type &value) const {
+    return value ? this->base_type::operator()(*value) : result_type();
+  }
+};
+template <>
+struct equal_to<shared_ptr<boost::process::child>> : equal_to<boost::process::child> {
+  typedef equal_to<boost::process::child> base_type;
+  typedef shared_ptr<boost::process::child> argument_type;
+  typedef bool result_type;
+  result_type operator()(const argument_type &a, const argument_type &b) const {
+    return a ? b && this->base_type::operator()(*a, *b) : !b;
+  }
+};
+template <>
+struct not_equal_to<shared_ptr<boost::process::child>>;
+template <>
+struct greater<shared_ptr<boost::process::child>>;
+template <>
+struct greater_equal<shared_ptr<boost::process::child>>;
+template <>
+struct less<shared_ptr<boost::process::child>> : less<boost::process::child> {
+  typedef less<boost::process::child> base_type;
+  typedef shared_ptr<boost::process::child> argument_type;
+  typedef bool result_type;
+  result_type operator()(const argument_type &a, const argument_type &b) const {
+    return a ? b && this->base_type::operator()(*a, *b) : !!b;
+  }
+};
+template <>
+struct less_equal<shared_ptr<boost::process::child>>;
+}  // namespace std
+
+namespace ray {
+typedef boost::process::pid_t pid_t;
```

Comment:
I don't know whether it's a good idea to do this `typedef`.
I'd like to rename this to some thing like `boost_pid_t` or `ray::Pid`.

Otherwise the developers will get confused once they `using ray;` in a cpp file.

Reply:
`using namespace NAMESPACE` isn't a good idea in the first place for this exact reason—you simply can't assume a namespace won't contain some conflicting name (and you naturally accept that risk if it does).

But I think the intention of this definition might be a bit different than what you're expecting. It's a very similar situation to e.g. `::size_t` and std::size_t`; there shouldn't really be any confusion. In fact, if I name it `Pid` now people are going to get confused as to what they should be using, whereas right now there's no distinction.

Note that this is intended for internal Ray usage, not for outsiders. And even within Ray, I'm trying to reduce the use of `pid_t` as much as possible (ideally to zero), not to suggest I'm trying to design an API around it. The `pid_t` here is only for compatibility to allow a smooth transition to that point, and to reduce dependence on `unistd.h` in the meantime, which is non-portable.

So it's fine :)